grammar.js

```diff
@@ -406,14 +406,47 @@ const rules = {
 
   // prettier-ignore
   string: $ =>
-    token(
-      choice(
-        /'(\\'|\\\\|\\?[^'\\])*'/,
-        /"(\\"|\\\\|\\?[^"\\])*"/,
-      ),
+    choice(
+      $._single_quoted_string,
+      $._double_quoted_string,
+    ),
+
+  // Breakout single quoted string for symmetry with double quoted strings.
+  _single_quoted_string: $ => /'(\\'|\\\\|\\?[^'\\])*'/,
+
+  // Reference: https://docs.hhvm.com/hack/source-code-fundamentals/literals#string-literals__double-quoted-string-literals
+  _double_quoted_string: $ =>
+    seq(
+      '"',
+      choice.rep($._string_character, $._escape_sequence, $.embedded_expression, $.embedded_brace_expression),
+      '"',
+    ),
+
+  _string_character: $ => choice(token(/([^"\\])/), token(prec(1, choice('#', '//', '/*')))),
```

Comment:
From last time we talked, did you try putting repeating `$._string_character` in it's own rule, so we can just have a `$.string_body` rule that encompass a full consecutive string?

```js
string_body: $ => rep1($._string_character)
```

or 

```js
string_body: $ => rep1(choice(token(/([^"\\])/), token(prec(1, choice('#', '//', '/*')))))
```

Reply:
Also, this is a nit, but I don't think you need `token` around `/([^"\\])/`. And that rule also doesn't have to use a regex group anymore `()` since there's only one option.

Reply:
When I try that it creates a conflict. Any suggestions for resolving it?

```
Unresolved conflict for symbol sequence:

  '"'  string_body_repeat1  •  '_string_character_token1'  …

Possible interpretations:

  1:  '"'  (string_body  string_body_repeat1)  •  '_string_character_token1'  …
  2:  '"'  (string_body_repeat1  string_body_repeat1  •  string_body_repeat1)

Possible resolutions:

  1:  Specify a left or right associativity in `string_body`
  2:  Add a conflict for these rules: `string_body`
```

I suspect that adding this rule as is might lead some problems. Tree-sitter prioritizes matching based on length. Say for example we have:

``
"sometext$var"
``

I think the `$` will be captured as part of the string body instead of an embedded expression, and therefore the whole thing will be parsed as a string body. We might need to make some modifications to accommodate this.

Reply:
I think you'd want right associativity for string_body since that'll capture the largest node. Left associativity would capture the string_body on first match so it'd end up just capturing individual characters.  

If we run into problems with `$` and `{`, we could do something like do for `$.xhp_comment` where we exclude problematic token characters and only allow them in specific sequences. Something like,

```js
  _comment_start_character: $ => token(prec(1, choice('#', '//', '/*'))),

  string_body: $ =>
    repeat1(
      choice(
        seq(opt('\\'), choice(/[^"\\${]/, $._comment_start_character)),
        seq(
          repeat1('$'),
          opt('\\'),
          choice(/[^"\\${a-zA-Z_\x80-\xff]/, $._comment_start_character),
        ),
        seq(
          repeat1('{'),
          opt('\\'),
          choice(/[^"\\${]/, $._comment_start_character),
        ),
      ),
    ),
```

But I just tried adding right associativity to repeating `$. _string_character` and it doesn't seem conflict with `$.variable` (honestly surprised it doesn't cause issues (at least not immediate ones)).